tests/gen/gen_predefined.go

```diff
@@ -0,0 +1,178 @@
+	resultDocs := make([]GeneratedDoc, 0, len(docsList.Docs))
+	typeDefs, err := parseSchema(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	generator := docGenerator{types: typeDefs}
+	for _, doc := range docsList.Docs {
+		docs, err := generator.generateRelatedDocs(doc, docsList.ColName)
+		if err != nil {
+			return nil, err
+		}
+		resultDocs = append(resultDocs, docs...)
+	}
+	return resultDocs, nil
+}
+
+// GeneratePredefined generates documents from a predefined list
+// of docs that might include nested docs.
+func GeneratePredefined(defs []client.CollectionDefinition, docsList DocsList) ([]GeneratedDoc, error) {
+	resultDocs := make([]GeneratedDoc, 0, len(docsList.Docs))
+	typeDefs := make(map[string]client.CollectionDefinition)
+	for _, col := range defs {
+		typeDefs[col.Description.Name] = col
+	}
+	generator := docGenerator{types: typeDefs}
+	for _, doc := range docsList.Docs {
+		docs, err := generator.generateRelatedDocs(doc, docsList.ColName)
+		if err != nil {
+			return nil, err
+		}
+		resultDocs = append(resultDocs, docs...)
+	}
+	return resultDocs, nil
+}
+
+type docGenerator struct {
+	types map[string]client.CollectionDefinition
+}
+
+// toRequestedDoc removes the fields that are not in the SDL of the collection.
+func toRequestedDoc(doc map[string]any, typeDef *client.CollectionDefinition) map[string]any {
+	result := make(map[string]any)
+	for _, field := range typeDef.Schema.Fields {
+		if field.IsRelation() || field.Name == request.KeyFieldName {
+			continue
+		}
+		result[field.Name] = doc[field.Name]
+	}
+	for name, val := range doc {
+		if strings.HasSuffix(name, request.RelatedObjectID) {
+			result[name] = val
+		}
+	}
+	return result
+}
+
+// generatePrimary generates primary docs for the given secondary doc.
+func (this *docGenerator) generatePrimary(
+	docMap map[string]any,
+	typeDef *client.CollectionDefinition,
+) (map[string]any, []GeneratedDoc, error) {
+	result := []GeneratedDoc{}
+	requested := toRequestedDoc(docMap, typeDef)
+	for _, field := range typeDef.Schema.Fields {
+		if field.IsRelation() {
+			if _, hasProp := docMap[field.Name]; hasProp {
+				if field.IsPrimaryRelation() {
+					subType := this.types[field.Schema]
+					subDocMap := toRequestedDoc(docMap[field.Name].(map[string]any), &subType)
+					subDoc, err := client.NewDocFromMap(subDocMap)
+					if err != nil {
+						return nil, nil, NewErrFailedToGenerateDoc(err)
+					}
+					requested[field.Name+request.RelatedObjectID] = subDoc.Key().String()
+					result = append(result, GeneratedDoc{Col: &subType, Doc: subDoc})
+				}
+			}
+		}
+	}
+	return requested, result, nil
+}
+
+// generateRelatedDocs generates related docs (primary and secondary) for the given doc.
+func (this *docGenerator) generateRelatedDocs(docMap map[string]any, typeName string) ([]GeneratedDoc, error) {
+	typeDef := this.types[typeName]
+
+	requested, result, err := this.generatePrimary(docMap, &typeDef)
```

Comment:
todo: This call must be made before doing anything else in this function due to the way `generatePrimary` mutates the document, however it is not obvious from reading the code.  I think this temporal coupling needs to be highlighted with a code comment, it would be easy enough to miss and then waste time on fixing should anyone start moving stuff around in this function.

Reply:
I don't understand your point. `generatePrimary` is called before anything else is done.

Reply:
Yes it is called before anything else is done, right now, in this version of the code.

However what is not clear from reading the code (besides the type change) is that it absolutely has to be called first because it mutates the document.

If anyone (including your later self) feels the need to change this function, it would be trivial to miss such a thing and then waste time trying to correct their mistake.